### DISCUSSION TITLE
The nested ProviderScope ignores its overrides and uses root overrides

### DISCUSSION THREAD

**Discussion by PavelPZ** [rrousselGit/riverpod#1175] [Q&A] [answered]
Following example raises error (see ERROR HERE bellow). I can't recognize it's a duplicate of that https://github.com/rrousselGit/river_pod/issues/669 or https://github.com/rrousselGit/river_pod/issues/575.

I don't understand how to implement your **"When scoping a provider, you should override its dependencies too"** advice in my case (I need the Ref in Navigator constructor).

```dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

void main() => runApp(
      ProviderScope(
        overrides: [
          navigatorCreatorProvider.overrideWithValue(RootNavigator.new),
        ],
        child: App(),
      ),
    );
typedef NavigatorCreator = Navigator Function(Ref);

abstract class Navigator {
  Navigator(this.ref);
  final Ref ref;
}

final navigatorCreatorProvider = Provider<NavigatorCreator>((_) => throw UnimplementedError());

final navigatorProvider = Provider<Navigator>((ref) {
  final creator = ref.read(navigatorCreatorProvider);
  return creator(ref);
});

class RootNavigator extends Navigator {
  RootNavigator(Ref ref) : super(ref);
}

class InnerNavigator extends Navigator {
  InnerNavigator(Ref ref) : super(ref);
}

class App extends ConsumerWidget {
  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final navigator = ref.read(navigatorProvider);
    assert(navigator is RootNavigator);
    return ProviderScope(
      overrides: [
        navigatorCreatorProvider.overrideWithValue(InnerNavigator.new),
      ],
      child: InnerApp(),
    );
  }
}

class InnerApp extends ConsumerWidget {
  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final navigator = ref.read(navigatorProvider);
   // ******************************************************************************* ERROR HERE
    assert(navigator is InnerNavigator); 
    return Text('Inner');
  }
}

```

**Comment by rrousselGit** [comment, marked as answer]
You're supposed to do:

```dart
    return ProviderScope(
      overrides: [
        navigatorCreatorProvider.overrideWithValue(InnerNavigator.new),
        navigatorProvider,
      ],
      child: InnerApp(),
    );
```

or:

```dart
final navigatorProvider = Provider<Navigator>((ref) {
  final creator = ref.read(navigatorCreatorProvider);
  return creator(ref);
}, dependencies: [navigatorCreatorProvider]);
```


Although that `NavigatorCreator` looks fishy to me.


You probably want to remove `navigatorCreatorProvider` and do:


```dart
    return ProviderScope(
      overrides: [
        navigatorProvider.overrideWithProvider(Provider((ref) => InnerNavigator(ref)),
      ],
      child: InnerApp(),
    );
```

**Reply by PavelPZ** [reply]
You saved me, thank you very much for your quick reply

**Comment by PavelPZ** [comment]
ignore this reply...

**Comment by ipcjs** [comment]
The function of the `dependencies` parameter is difficult to understand. 😵

Maybe it should be specifically stated in the documentation.

**Comment by dgaedcke** [comment]
I'm using straight RP (not Flutter) in a Dart only project ... and it seems clear to me that WHEN there is a "parent provider container", the child-override is being missed, EVEN WHEN the "ref.read" originates with the child (overridden) ProviderContainer ... my example is too complex to share, but if this doesn't tweak your intuition, let me know and I'll try to find time to build a simple proof ...

**Reply by rrousselGit** [reply]
If you're manually creating ProviderContainers, you need to specify their "parent" parameter if you want to "scope" something

**Reply by dgaedcke** [reply]
I don't think I expressed myself very clearly:
When I send null to the parent argument, my ref.read() succeeds (properly respects the override)
When I pass an actual ProviderContainer to the parent argument, the EXACT SAME ref.read() code fails by hitting the throw Exception in my provider definition ... in other words, it doesn't see or respect my override!!

**Reply by rrousselGit** [reply]
I assume that `ref.read` is from a provider?

Chances are you need to specify the `dependencies` parameter

**Reply by dgaedcke** [reply]
yes, most of the time, we inject ref into the model object at creation time.
In a few very limited exceptions, we use providerContainer.read()

I tried to prove this was broken here:
https://dartpad.dev/?id=c791d537fc61cff9b0ec82557f2da37d

but it seems I've failed.  And it seems dartpad won't let me use straight riverpod ... it's demanding the flutter_riverpod package instead.
But given that failure, I'm going to conclude that I've not set all of my dependencies properly and go back and try again.

Do you have any docs on the dependencies argument that you can share??

**Reply by rrousselGit** [reply]
I'm not quite understanding your problem. With the way you've coded things, both your `testCaseShouldFail` and `testCaseShouldWork` correctly override `_ncInternalStateProvider`, so neither of them reach the unimplementederror

What's the issue here?

**Reply by dgaedcke** [reply]
in our actual code, the case where (parent: null)  DOES WORK ... does not throw the error!
Only in the case where we pass an ACTUAL parent PC, do we hit the (missed override) error

In other words, the CHILD override seems to be missed WHEN THERE IS A PARENT scope ...

I was expecting the same failure here, and it's not happening, so I'm guessing it's an issue with our dependencies list ...

Do you have any docs on the dependencies argument that you can share??

**Reply by rrousselGit** [reply]
I'm missing something here. In your dartpad, neither of the functions throw.

With the code you gave, `dependencies` doesn't have an effect. I don't see an issue with your code anyway, so I'm confused.

**Reply by dgaedcke** [reply]
Yes, I keep saying that my dartpad has FAILED TO DEMONSTRATE the error we're seeing in live code.
So I'm not able to reproduce the problem in a SIMPLE case.
That's why you are not seeing my issue in code ... but you should be able to see that I've explained that it fails when we pass a "parent" argument to the child PC ... 

So now I want to go back to the complex case (in which we DO PASS a parent PC arg) and experiment with fixing it.

To do so, I need to understand if "dependencies" should include:
a) providers I read
b) providers I watch
c)  providers in either / both categories

And further, should dependencies ONLY include things I watch or read:
1)  IN THE CONSTRUCTOR of the provider
or does it include providers
2)  I read inside the object/class using my stored "ref" argument??

I'm trying really hard to communicate here and I'm very confused about why I'm failing ... I hope this fully explains our situation!!

**Reply by rrousselGit** [reply]
Looking at your dartpad, my guess is that the problem isn't with a missing `dependencies`. Rather, maybe you used the wrong ProviderContainer

In your dartpad, if you did:

```dart
  var person = parentPC.read(_ncInternalStateProvider);
```

instead of:

```dart
  var person = childPC.read(_ncInternalStateProvider);
```

then you'd get an exception, as _ncInternalStateProvider isn't overridden on parentPC

**Reply by rrousselGit** [reply]
As for `dependencies`, that should be the list of all the providers that this provider uses. Be it using read, watch, listen or anything really (even `invalidate`, ...)

**Reply by dgaedcke** [reply]
yes, that is the exception we get!!
And yes, I had assumed we were trying to read using parentPC ... but I've confirmed that is NOT the case ... it's the exact same ref.read() that fails or works, depending on whether we have passed a parent or not ... the rest of the code remains unchanged ... so other than dependencies, I'm not sure where else to look ...

thanks for the docs!

**Reply by rrousselGit** [reply]
Without a way to reproduce what you're describing, it's hard for me to help.

It'd be great if you could isolate the issue and extract it into dartpad as you tried before

**Reply by dgaedcke** [reply]
Progress!!!  Now that all the dependencies are fully specified, I'm now seeing a new error:
`ui_dart_state.cc(198)] Unhandled Exception: 'package:riverpod/src/framework/provider_base.dart': Failed assertion: line 617 pos 7: 'this.provider != origin ||
package:riverpod/…/framework/provider_base.dart:617
          origin.dependencies == null ||
          origin.dependencies!.contains(provider.from) ||
          origin.dependencies!.contains(provider)': The provider Provider<FrontBackGroundObserver>#800b0 tried to read fgbgStreamProvider.stream:AsyncValueAsStreamProvider<FGBGType>#58c5f, but it specified a 'dependendencies' list yet that list does not contain fgbgStreamProvider.stream:AsyncValueAsStreamProvider<FGBGType>#58c5f.
To fix, add fgbgStreamProvider.stream:AsyncValueAsStreamProvider<FGBGType>#58c5f to Provider<FrontBackGroundObserver>#800b0's 'dependencies' parameter
#0      _AssertionError._doThrowNew (dart:core-patch/errors_patch.dart:51:61)
#1      _AssertionError._throwNew (dart:core-patch/errors_patch.dart:40:5)
#2      ProviderElementBase._debugAssertCanDependOn (package:riverpod/src/framework/provider_base.dart:617:7)<…>`

**Reply by rrousselGit** [reply]
This error is telling you that you didn't specify `dependencies` correctly and you're missing some providers

Currently you need to specify the `.stream` & co in the list too. That requirement will not be necessary in the future though.

**Reply by dgaedcke** [reply]
Got it!!  thank you ...  now that all dependencies are specified, the problems have gone away ... 

It now works just as well as it does when we don't pass a parent PC!!

Thanks for your help!!  Really great library :-)
